include/core/orderedjson.h

```diff
@@ -0,0 +1,316 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#pragma once
+
+#include <QString>
+#include <QVector>
+#include <QMap>
+#include <QPair>
+#include <QFile>
+#include <QTextStream>
+
+#include <memory>
+#include <initializer_list>
+
+#ifdef _MSC_VER
+    #if _MSC_VER <= 1800 // VS 2013
+        #ifndef noexcept
+            #define noexcept throw()
+        #endif
+
+        #ifndef snprintf
+            #define snprintf _snprintf_s
+        #endif
+    #endif
+#endif
+
+namespace poryjson {
+
+template <class T>
+class OrderedMap : public QMap<QString, T> {
+    //
+public:
+    OrderedMap() {}
+    ~OrderedMap() {}
+
+    // key value iterator
+    class iterator {
+    public:
+        iterator(const OrderedMap *map, int pos) : m_map(map), m_pos(pos) {}
+
+        bool operator != (const iterator& other) const {
+            return m_pos != other.m_pos;
+        }
+
+        bool operator == (const iterator& other) const {
+            return m_pos == other.m_pos;
+        }
+
+        iterator & operator = (const iterator &other) {
+            this->m_pos = other.m_pos;
+            this->m_map = other.m_map;
+        }
+
+        T & value () const {
+            return (*m_map)[m_map->m_keyOrder.value(m_pos)];
+        }
+
+        QPair<QString, T> operator * () const {
+            QString key = m_map->m_keyOrder.value(m_pos);
+            return QPair<QString, T>(key, m_map->value(key));
+        }
+
+        const iterator & operator ++ () {
+            m_pos++;
+            return *this;
+        }
+
+    private:
+        const OrderedMap *m_map;
+        int m_pos;
+    };
+
+    const iterator begin() const { return iterator(this, 0); }
+
+    const iterator end() const { return iterator(this, this->size()); }
+
+    iterator find(const QString &key) const {
+        int pos = 0;
+        for (auto i : *this) {
+            if (i.first == key) {
+                return iterator(this, pos);
+            }
+        }
+        return end();
+    }
+
+    bool operator < (const OrderedMap<T>& other) const {
+        return m_keyOrder.size() < other.m_keyOrder.size();
+    }
+
+    // only works for inserting unique keys!
+    T & operator [] (const QString &key) {
```

Comment:
I don't understand what functionality this method is overriding.  Why is it inserting a default value into the underlying `QMap`, rather than an actual value?

Reply:
It appears to be this function? https://doc.qt.io/qt-5/qmap.html#operator-5b-5d

If that is true, then this isn't implemented correctly, since it's supposed to insert only if there is no associated item with `key`.

Reply:
I'm further confused because this is the only place `m_keyOrder` is appended to, but I would expect you to have overridden `QMap`'s `insert` method, too, and updated `m_keyOrder` there.

Reply:
Well, it's this way because for porymap we don't try to insert keys that aren't unique or reassign values, and they are only ever inserted with `operator []`. I wanted to keep this as small as possible, maybe not the best idea? When I started this, for some reason I thought Qt already had an ordered map class and was planning to use that. 
I can update this so that it follows a more standard behavior.
Another option is to just use another MIT licensed library that already has all these problems solved. (Or I was thinking I could modify QMap by copying the code from Qt's source, but I do not fully understand the license permissions although I'm pretty sure that's allowed).

Reply:
I want to avoid a situation where unexpected things happen, and this `QMap`-override implementation has some holes that could potentially cause headaches down the road (the risk is admittedly small).  I'd prefer to use https://github.com/Tessil/ordered-map for the `OrderedMap` implementation--it's just a drop-in header, so I wouldn't expect it to be too much work to switch over.

Reply:
I agree this is the safer option. Would you rather have these things in include/core/ or like an include/external for organization purposes?